A partir del dataset mnist (tf.keras.datasets.mnist.load_data()), realiza una clasificación usando:

1. Una CNN con:
    - 1 capa convolutiva con 8 neuronas
    - 1 MaxPool quedando las dimensiones de la imagen a la mitad
    - 1 dropout 0.25
    - 1 Flatten
    - 1 dense con 32 neuronas
    - 1 dense con 10 (salida)

2. Una CNN con:
    - 1 capa convolutiva con 8 neuronas
    - 1 MaxPool quedando las dimensiones de la imagen a la mitad
    - 1 dropout 0.25
    - 1 Flatten
    - 1 dense con 16 neuronas
    - 1 dense con 32 neuronas
    - 1 dense con 10 (salida)

¿ Cuál ha dado mejor resultado?

Para compilar el modelo, usa:

In [ ]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [2]:
# TensorFlow y tf.keras
import tensorflow as tf
from tensorflow import keras

# Librerias de ayuda
import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.5.0


In [3]:
fashion_mnist = keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

In [8]:
train_images.max()

255

In [3]:
train_images = train_images / 255.0
test_images = test_images / 255.0

In [4]:
train_images.shape

(60000, 28, 28)

In [20]:
train_images_2 = train_images.reshape(60000,28,28,1)
train_images_2.shape

(60000, 28, 28, 1)

### Una CNN con:

- 1 capa convolutiva con 8 neuronas
- 1 MaxPool quedando las dimensiones de la imagen a la mitad
- 1 dropout 0.25
- 1 Flatten
- 1 dense con 32 neuronas
- 1 dense con 10 (salida)

In [6]:
model = keras.Sequential([
    keras.layers.Conv2D( filters=8, kernel_size=(3, 3), padding="same", input_shape=(28, 28, 1),activation='relu'),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Dropout(0.25),
    keras.layers.Flatten(),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(10, activation="softmax")
])

### Compile el modelo

In [7]:
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [8]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 28, 28, 8)         80        
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 14, 14, 8)         0         
_________________________________________________________________
dropout (Dropout)            (None, 14, 14, 8)         0         
_________________________________________________________________
flatten (Flatten)            (None, 1568)              0         
_________________________________________________________________
dense (Dense)                (None, 32)                50208     
_________________________________________________________________
dense_1 (Dense)              (None, 10)                330       
Total params: 50,618
Trainable params: 50,618
Non-trainable params: 0
____________________________________________________

### Entrenar el Modelo

In [9]:
model.fit(train_images_2, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.5024 - accuracy: 0.8229
Epoch 2/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.3579 - accuracy: 0.8714
Epoch 3/10
1875/1875 [==============================] - 16s 8ms/step - loss: 0.3213 - accuracy: 0.8849
Epoch 4/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.3004 - accuracy: 0.8913
Epoch 5/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2831 - accuracy: 0.8970
Epoch 6/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.2720 - accuracy: 0.9012
Epoch 7/10
1875/1875 [==============================] - 12s 7ms/step - loss: 0.2623 - accuracy: 0.9040
Epoch 8/10
1875/1875 [==============================] - 12s 7ms/step - loss: 0.2515 - accuracy: 0.9070
Epoch 9/10
1875/1875 [==============================] - 12s 7ms/step - loss: 0.2446 - accuracy: 0.9104
Epoch 10/10
1875/1875 [==============================] - 12s 7ms/step - l

A medida que el modelo entrena, la perdida y la exactitud son desplegadas. Este modelo alcanza una exactitud de 0.91 (o 91%) sobre el set de datos de entrenamiento.

### Evaluar Exactitud

In [18]:
test_images_2 = test_images.reshape(10000,28,28,1)
test_images_2.shape

(10000, 28, 28, 1)

In [12]:
test_loss, test_acc = model.evaluate(test_images_2,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

313/313 - 1s - loss: 0.2701 - accuracy: 0.9018

Test accuracy: 0.9017999768257141


La exactitud sobre el set de datos es similar, considerandose una diferencia minima de 0.01 sobre el set de entrenamiento.

2. Una CNN con:
    - 1 capa convolutiva con 8 neuronas
    - 1 MaxPool quedando las dimensiones de la imagen a la mitad
    - 1 dropout 0.25
    - 1 Flatten
    - 1 dense con 16 neuronas
    - 1 dense con 32 neuronas
    - 1 dense con 10 (salida)

In [14]:
model_2 = keras.Sequential([
    keras.layers.Conv2D( filters=8, kernel_size=(3, 3), padding="same", input_shape=(28, 28, 1),activation='relu'),
    keras.layers.MaxPooling2D((2, 2)),
    keras.layers.Dropout(0.25),
    keras.layers.Flatten(),
    keras.layers.Dense(16, activation='relu'),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(10, activation="softmax")
])

In [15]:
model_2.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [16]:
model_2.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 28, 28, 8)         80        
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 14, 14, 8)         0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 14, 14, 8)         0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 1568)              0         
_________________________________________________________________
dense_3 (Dense)              (None, 16)                25104     
_________________________________________________________________
dense_4 (Dense)              (None, 32)                544       
_________________________________________________________________
dense_5 (Dense)              (None, 10)               

### Entrenar el Modelo

In [21]:
model_2.fit(train_images_2, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.9884 - accuracy: 0.6458
Epoch 2/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.4735 - accuracy: 0.8342
Epoch 3/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.4165 - accuracy: 0.8543
Epoch 4/10
1875/1875 [==============================] - 20s 11ms/step - loss: 0.3872 - accuracy: 0.8626
Epoch 5/10
1875/1875 [==============================] - 16s 9ms/step - loss: 0.3619 - accuracy: 0.8717
Epoch 6/10
1875/1875 [==============================] - 17s 9ms/step - loss: 0.3477 - accuracy: 0.8759
Epoch 7/10
1875/1875 [==============================] - 15s 8ms/step - loss: 0.3359 - accuracy: 0.8796
Epoch 8/10
1875/1875 [==============================] - 13s 7ms/step - loss: 0.3251 - accuracy: 0.8832
Epoch 9/10
1875/1875 [==============================] - 15s 8ms/step - loss: 0.3202 - accuracy: 0.8841
Epoch 10/10
1875/1875 [==============================] - 13s 7ms/step - 

#### Los datos de entrenamiento alcanzan una exactitud del 0.89 (89%)

### Evaluar Exactitud

In [22]:
test_loss, test_acc = model_2.evaluate(test_images_2,  test_labels, verbose=2)

313/313 - 1s - loss: 0.3529 - accuracy: 0.8768


#### Los datos de evaluacion alcanzan una exactitud del 0.90% es decir, un resultado similar a los datos de entrenamiento 

### Ambos modelos han dado buenos resultados, sin embargo el modelo anterior presenta un resultado mayor.